relay-metrics/src/aggregator.rs

```diff
@@ -16,6 +17,275 @@ use tokio::time::Instant;
+                    aggregator = agg.name(),
+                );
+            }
+        }
+    }
+}
+
+/// Splits this bucket if its estimated serialization size exceeds a threshold.
+///
+/// There are three possible return values:
+///  - `(Some, None)` if the bucket fits entirely into the size budget. There is no split.
+///  - `(None, Some)` if the size budget cannot even hold the bucket name and tags. There is no
+///    split, the entire bucket is moved.
+///  - `(Some, Some)` if the bucket fits partially. Remaining values are moved into a new bucket
+///    with all other information cloned.
+///
+/// This is an approximate function. The bucket is not actually serialized, but rather its
+/// footprint is estimated through the number of data points contained. See
+/// `estimate_size` for more information.
+fn split_at(mut bucket: Bucket, size: usize) -> (Option<Bucket>, Option<Bucket>) {
+    // If there's enough space for the entire bucket, do not perform a split.
+    if size >= estimate_size(&bucket) {
+        return (Some(bucket), None);
+    }
+
+    // If the bucket key can't even fit into the remaining length, move the entire bucket into
+    // the right-hand side.
+    let own_size = estimate_base_size(&bucket);
+    if size < (own_size + AVG_VALUE_SIZE) {
+        // split_at must not be zero
+        return (None, Some(bucket));
+    }
+
+    // Perform a split with the remaining space after adding the key. We assume an average
+    // length of 8 bytes per value and compute the number of items fitting into the left side.
+    let split_at = (size - own_size) / AVG_VALUE_SIZE;
+
+    match bucket.value {
+        BucketValue::Counter(_) => (None, Some(bucket)),
+        BucketValue::Distribution(ref mut distribution) => {
+            let mut org = std::mem::take(distribution);
+
+            let mut new_bucket = bucket.clone();
+            new_bucket.value =
+                BucketValue::Distribution(DistributionValue::from_slice(&org[split_at..]));
+
+            org.truncate(split_at);
+            bucket.value = BucketValue::Distribution(org);
+
+            (Some(bucket), Some(new_bucket))
+        }
+        BucketValue::Set(ref mut set) => {
+            let org = std::mem::take(set);
+            let mut new_bucket = bucket.clone();
+
+            let mut iter = org.into_iter();
+            bucket.value = BucketValue::Set((&mut iter).take(split_at).collect());
+            new_bucket.value = BucketValue::Set(iter.collect());
+
+            (Some(bucket), Some(new_bucket))
+        }
+        BucketValue::Gauge(_) => (None, Some(bucket)),
+    }
+}
+
+/// Estimates the number of bytes needed to serialize the bucket without value.
+///
+/// Note that this does not match the exact size of the serialized payload. Instead, the size is
+/// approximated through tags and a static overhead.
+fn estimate_base_size(bucket: &Bucket) -> usize {
+    50 + bucket.name.len() + aggregator::tags_cost(&bucket.tags)
```

Comment:
Can we move this `50` in the constant with the name instead?

Reply:
you mean adding a new constant? I did that now. 

note: I simply moved this function, I did not create it.